src/Orleans/Utils/IKeyedServiceCollection.cs

```diff
@@ -10,7 +12,40 @@ namespace Orleans.Runtime
     public interface IKeyedServiceCollection<in TKey, out TService>
         where TService : class
     {
-        TService GetService(TKey key);
+        TService GetService(IServiceProvider services, TKey key);
+    }
+
+    public interface IKeyedService<TKey, out TService> : IEquatable<TKey>
+    {
+        TService GetService(IServiceProvider services);
+    }
+
+    public class KeyedService<TKey, TService, TInstance> : IKeyedService<TKey, TService>
+            where TInstance : TService
+    {
+        private readonly TKey key;
+
+        public KeyedService(TKey key)
+        {
+            this.key = key;
+        }
+
+        public TService GetService(IServiceProvider services) => services.GetService<TInstance>();
```

Comment:
> Less convinced I am that we should be putting the name mapping itself as singleton services in DI.

Understood.  I'm not committed to this solution.  We discussed solving this different ways (the prototype you linked, earlier factory patterns, possibly using named options, ..., but none of those are in place now.  If/when we spend the time to improve our support for named services, I'd expect this code to be refactored/replaced/improved.

> There is no need to have the factory be a transient service

It's worse than that.  It's a scoped service.  It needs to be that way currently to obtain the context or the correct service provider.  I spoke with Sergey about this problem and will be opening an issue about this, I simply didn't want to complicate this PR with that problem, as it is something we can address in future changes.  To be clear, the requirement that the factories be scoped is orthogonal to the named mappings being in DI.  These are two separate (separable) problems.

> If you want, I can submit a small PR adapting that old branch to your branch specifically.

I would love to see improvements on this, after this PR is checked in and unblocks other work.  This has always been the plan.  I'm less open to introducing further changes at this stage that do not address specific problems in the general pattern.  The use of the keyed services is entirely within the ExampleStorage sample code and is not part of the general pattern.  Features can do this any way they like, as long as it works with the framework and respects our DI limitations.  I'd, however, object to each feature providing it's own keyed service implementation, as that would incur a high maintenance cost.  In short, we should solve the keyed service problem and use the same solution for all our feature factories, but that does not mean the solution in this pr **IS** that solution.

> have the INamedExampleStorageFactory receive the IGrainActivationContext on every call to Create.

I am not 100% against this, but the feature factories (INamedExampleStorageFactory, IExampleStorageFactory) are also intended to be injected directly into the grain and called by the grain logic, which does not have access to the IGrainActivationContext, unless it too is passed into the grain constructor.  There are ways around this, however, one of the requirements negotiated in the general pattern was that the attribute handling logic must use the same factories that the grain does to create the feature, so if we add the IGrainActivationContext to the factories, it must be passed by the grains, if we don't want to pass the grain context in the grains, the attribution logic can't pass it either (or we abandon that negotiated requirement).